### DISCUSSION TITLE
Single repository for gcode via GDrive/SMBNSF/NAS

### DISCUSSION THREAD

**Discussion by Firefighter1023** [OctoFarm/OctoFarm#1264] [Feature Request]
I maintain a repository of all my gcode, typically in gdrive, where from this location I upload the gcode to each octoprint instance for execution.  The gcode files are organized in multi folders and sub folders.  At this time, uploading an entire directory tree, and maintaining the same folder/subfolder directory to every octoprint instance isn't possible.

It seems beneficial for Octofarm to have the ability to "scrape" a network directory, and present the results of that scrape in the octofarm   GUI.  The user can select which gcode to print.

Having a central repository can certainly help ensure that the operator is constantly using the most up to date gcode files, and eliminate the need to upload updated gcode to multiple octoprint instances, or executing old gcode which may be misconfigured.

It is preferred to be able to use gdrive as a source, however other mainstream network data sources would be good to be implemented for flexibility.

Thank you for the awesome software, it's truly a life saver!

**Comment by NotExpectedYet** [comment]
Ah we were talking about ever so slightly different things aha. Doesn't really make much difference the end result is the same. 

I was going to allow OctoFarm to manage a local file system folder structure. It can create the repo for you then, and would support all you listed because you could just sync your gdrive in that folder or setup a SMB share on the other side. My way is easier on me, and more on the user 😂. 

Both have Merritt. 

I think I might at least start laying the ground work for these features though. Currently the multi-upload and multi-print options for example all get handled client side. Time to push those server side.

**Comment by Firefighter1023** [comment]
I do agree, I think your implementation would be easy and accomplish the same goal.  I do appreciate your input and consideration on this!

**Reply by NotExpectedYet** [reply]
Your most welcome. 

Talking the mult-upload to the server will at least get me part way there to supporting uploads without the interface been needed. 

Once that's done, I believe, after looking it up more, that I can support uploading a folder, without it zipping and possible keeping the structure. Certainly seems possible in browsers now for top level, will have to test further see if it will replicate the entire structure. 

Dropbox does it already so it's certainly possible. 

I'm hoping for it to work a little like this. 

Drop, or select folder into file upload. Make a note of printers those files are getting replicated on. Pass all the files up to the server, the server will then keep the files in a local cache whilst it simultaneously uploads to multiple printers at once. So you'd get 1 file replicated on the required printer list, then it would move onto the next file. In my head that should work anyway ahaha.
